src/Compilers/CSharp/Portable/Binder/LocalScopeBinder.cs

```diff
@@ -181,6 +189,34 @@ protected ImmutableArray<LocalSymbol> BuildLocals(SyntaxList<StatementSyntax> st
             return locals?.ToImmutableAndFree() ?? ImmutableArray<LocalSymbol>.Empty;
         }
 
+        // When a VariableDeclaration is used in a deconstruction, there are two cases:
+        // - deconstruction is set, type may be set (for "var"), and no declarators. For instance, `var (x, ...)` or `(int x, ...)`.
+        // - deconstruction is null, type may be set, and there is one declarator holding the identifier. For instance, `int x` or `x`.
+        private void CollectLocalsFromDeconstruction(VariableDeclarationSyntax declaration, TypeSyntax varSyntax, ArrayBuilder<LocalSymbol> locals)
+        {
+            if (declaration.Deconstruction != null)
+            {
+                var nestedVarSyntax = varSyntax ?? declaration.Type;
+                foreach (var variable in declaration.Deconstruction.Variables)
+                {
+                    CollectLocalsFromDeconstruction(variable, nestedVarSyntax, locals);
+                }
+            }
+            else
+            {
+                Debug.Assert(declaration.Variables.Count == 1);
+                var declarator = declaration.Variables[0];
+
+                SourceLocalSymbol localSymbol = SourceLocalSymbol.MakeDeconstructionLocal(
+                                                            this.ContainingMemberOrLambda,
+                                                            this,
+                                                            declaration.Type,
+                                                            declarator.Identifier,
+                                                            isVar: varSyntax != null);
```

Comment:
> it cannot be observed

Even if that is the case today. Judging about different "properties" of the type based on the same syntax makes code easier to reason about, less error prone, and more robust against future changes. For example we might relax parser behavior in error recovery to allow types here, even though it would cause parse errors.

---
In reply to: [68824220](https://github.com/dotnet/roslyn/pull/12224#discussion_r68824220) [](ancestors = 68824220)